conan/cache/conan_reference_layout.py

```diff
@@ -0,0 +1,181 @@
+import os
+from contextlib import contextmanager
+
+from conan.cache.conan_reference import ConanReference
+from conans.errors import ConanException
+from conans.model.manifest import FileTreeManifest
+from conans.paths import BUILD_FOLDER, PACKAGES_FOLDER, SYSTEM_REQS_FOLDER, SYSTEM_REQS, DATA_YML, \
+    rm_conandir, EXPORT_SRC_FOLDER, SRC_FOLDER
+from conans.paths import CONANFILE, SCM_SRC_FOLDER
+from conans.util.files import rmdir
+from conans.util.files import set_dirty, clean_dirty, is_dirty
+
+
+class ReferenceLayout:
+    def __init__(self, ref, cache, base_folder):
+        self._ref = ref
+        self._cache = cache
+        self._base_folder = base_folder
+
+    @property
+    def reference(self):
+        if self._ref.pkgid:
+            return self._ref.as_package_reference()
+        else:
+            return self._ref.as_conanfile_reference()
+
+    def assign_prev(self, ref: ConanReference):
+        assert ref.reference == self._ref.reference, "You cannot change the reference here"
+        assert ref.prev, "It only makes sense to change if you are providing a package revision"
+        assert ref.pkgid, "It only makes sense to change if you are providing a package id"
+
+        old_pref = self._ref
+        self._ref = ref
+
+        new_path = self._cache._move_prev(old_pref, self._ref)
+        if new_path:
+            self._base_folder = new_path
+
+    def assign_rrev(self, ref: ConanReference):
+        assert ref.reference == self._ref.reference, "You cannot change reference name here"
+        assert ref.rrev, "It only makes sense to change if you are providing a revision"
+        assert not ref.prev, "The reference for the recipe should not have package revision"
+        assert not ref.pkgid, "The reference for the recipe should not have package id"
+
+        # TODO: here maybe we should block the recipe and all the packages too
+        old_ref = self._ref
+        self._ref = ref
+
+        # Move temporal folder contents to final folder
+        new_path = self._cache._move_rrev(old_ref, self._ref)
+        if new_path:
+            self._base_folder = new_path
+
+    @property
+    def base_folder(self):
+        return os.path.join(self._cache.base_folder, self._base_folder)
+
+    def build(self):
+        assert self._ref.pkgid, "Must be a reference of a package"
```

Comment:
I am not fully convinced about this, it might be better to split the class (same creation interface, transparent instantiation)